experimental/ast/codec.go

```diff
@@ -0,0 +1,469 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+import (
+	"fmt"
+	"reflect"
+	"slices"
+
+	"github.com/bufbuild/protocompile/experimental/internal"
+	"github.com/bufbuild/protocompile/experimental/report"
+	compilerpb "github.com/bufbuild/protocompile/internal/gen/buf/compiler/v1alpha1"
+)
+
+// codec is the state needed for converting an AST node into a Protobuf message.
+type codec struct {
+	*ToProtoOptions
+
+	stack    []report.Spanner
+	stackMap map[report.Spanner]bool
```

Comment:
For things that are really sets, I prefer `map[report.Spanner]struct{}`. Usage changes from:
```go
cycle = c.stackMap[v]
```
to this
```go
_, cycle = c.stackMap[v]
```
So I think it's just as clear/maintainable and elides unused space in the map.

Reply:
Funny, I used to use `struct{}`, because it is more unambiguously correct, but at some point I stopped caring.